File: fundref.ts
/**
 * this module provides normalization and validation of FundRefID from CrossRef
 * @module
 */
const fundRefPattern = "^10\\.\\d{4,9}\\/[-._;()/:A-Z0-9]+$";
const reFundRef = new RegExp(fundRefPattern, "i");

/**
 * normalizeFundRefID
 * @param id, FundRef identifier
 * @returns string
 * 
 * @example
 * ```ts
 * const fundRefID:string = nomralizeFundRefID('10.13039/100000001');
 * console.log(fundRefID); // 10.13039/100000001
 * ```
 */
export function normalizeFundRefID(id: string): string {
  return id.trim().toLowerCase();
}

/**
 * validateFundRefID
 * @param id, FundRef identifier
 * @returns true if valid, false otherwise
 * 
 * @example
 * ```ts
 * const fundRefID:string = nomralizeFundRefID('10.13039/100000001');
 * 
 * if (await validateFundRefID(fundRefID)) {
 *     console.log(`{fundRefID} appears valid`);
 * } else {
 *     console.error(`{fundRefID}, failed validation`);
 * }
 * ```
 */
export function validateFundRefID(id: string): boolean {
  const normalizedID = normalizeFundRefID(id);
  return reFundRef.test(normalizedID);
}

File: fundref.go
package metadatatools

import (
	"regexp"
	"strings"
)

// fundRefPattern matches a normalized FundRef DOI issued by CrossRef.
const fundRefPattern = `^10\.\d{4,9}/[-._;()/:A-Z0-9]+$`

// reFundRef is the compiled, case-insensitive form of fundRefPattern.
var reFundRef = regexp.MustCompile(`(?i)` + fundRefPattern)

// NormalizeFundRefID trims whitespace from id and lower cases it.
//
// Example:
//
//	id := NormalizeFundRefID("10.13039/100000001")
//	// id == "10.13039/100000001"
func NormalizeFundRefID(id string) string {
	return strings.ToLower(strings.TrimSpace(id))
}

// ValidateFundRefID normalizes id then validates it against fundRefPattern.
func ValidateFundRefID(id string) bool {
	return reFundRef.MatchString(NormalizeFundRefID(id))
}
